src/dns_server.c

```diff
@@ -1,95 +1,78 @@
-REQUEST_MAX = 1500  // A default MTU. We don't do TCP so any bigger is likely a waste
-};
-
-
 // Creates and bind a listening UDP socket for incoming requests.
-static int get_listen_sock(const char *listen_addr, int listen_port,
-                           unsigned int *addrlen) {
-  struct addrinfo *ai = NULL;
-  struct addrinfo hints;
-  // NOLINTNEXTLINE(clang-analyzer-security.insecureAPI.DeprecatedOrUnsafeBufferHandling)
-  memset(&hints, 0, sizeof(struct addrinfo));
-  /* prevent DNS lookups if leakage is our worry */
-  hints.ai_flags = AI_NUMERICHOST;
-
-  int res = getaddrinfo(listen_addr, NULL, &hints, &ai);
-  if(res != 0) {
-    FLOG("Error parsing listen address %s:%d (getaddrinfo): %s", listen_addr, listen_port,
-          gai_strerror(res));
-    if(ai) {
-      freeaddrinfo(ai);
-    }
-    return -1;
+static int get_listen_sock(struct addrinfo *listen_addrinfo) {
+  int sock = socket(listen_addrinfo->ai_family, SOCK_DGRAM, 0);
+  if (sock < 0) {
+    FLOG("Error creating socket: %s (%d)", strerror(errno), errno);
   }
 
-  struct sockaddr_in *saddr = (struct sockaddr_in*) ai->ai_addr;
-
-  *addrlen = ai->ai_addrlen;
-  saddr->sin_port = htons(listen_port);
-
-  int sock = socket(ai->ai_family, SOCK_DGRAM, 0);
-  if (sock < 0) {
-    FLOG("Error creating socket");
+  uint16_t port;
+  char ipstr[INET6_ADDRSTRLEN];
+  if (listen_addrinfo->ai_family == AF_INET) {
+    port = ntohs(((struct sockaddr_in*) listen_addrinfo->ai_addr)->sin_port);
+    inet_ntop(AF_INET, &((struct sockaddr_in *)listen_addrinfo->ai_addr)->sin_addr, ipstr, sizeof(ipstr));
+  } else if (listen_addrinfo->ai_family == AF_INET6) {
+    port = ntohs(((struct sockaddr_in6*) listen_addrinfo->ai_addr)->sin6_port);
+    inet_ntop(AF_INET6, &((struct sockaddr_in6 *)listen_addrinfo->ai_addr)->sin6_addr, ipstr, sizeof(ipstr));
+  } else {
+    FLOG("Unknown address family: %d", listen_addrinfo->ai_family);
   }
 
-  res = bind(sock, ai->ai_addr, ai->ai_addrlen);
+  int res = bind(sock, listen_addrinfo->ai_addr, listen_addrinfo->ai_addrlen);
   if (res < 0) {
-    FLOG("Error binding %s:%d: %s (%d)", listen_addr, listen_port,
-         strerror(errno), res);
+    FLOG("Error binding on %s:%d UDP: %s (%d)", ipstr, port,
+         strerror(errno), errno);
   }
 
-  freeaddrinfo(ai);
+  ILOG("Listening on %s:%d UDP", ipstr, port);
 
-  ILOG("Listening on %s:%d", listen_addr, listen_port);
   return sock;
 }
 
 static void watcher_cb(struct ev_loop __attribute__((unused)) *loop,
                        ev_io *w, int __attribute__((unused)) revents) {
   dns_server_t *d = (dns_server_t *)w->data;
 
-  char *buf = (char *)calloc(1, REQUEST_MAX + 1);
-  if (buf == NULL) {
-    FLOG("Out of mem");
-  }
-  struct sockaddr_storage raddr;
-  /* recvfrom can write to addrlen */
-  socklen_t tmp_addrlen = d->addrlen;
-  ssize_t len = recvfrom(w->fd, buf, REQUEST_MAX, 0, (struct sockaddr*)&raddr,
+  char tmp_buf[UINT16_MAX];  // stack buffer for largest UDP packet to support EDNS
```

Comment:
I worry a little about this. It's probably OK but might blow up on some small systems with limited stack space.
We copy it into a heap allocated buffer anyway, so what about 
```
char *dns_req = (char *)malloc(UINT16_MAX);
recvfrom(...);
...
dns_req = (char *)realloc(dns_req, len);
if (dns_req == NULL) { ... }
```

Reply:
Even on openwrt the stack limit is 8MB per thread. Stack size grows only by usage. For my taste 64k is not a huge deal and I would not mess with heap for temporary buffers. Also I would not reserve more on heap (and shrik later) to avoid heap fragmentation. So I'm not expecting any issues with this.
Anyway: I will think about this.